### ISSUE TITLE
Improve detection of simple comma-separated lists in query string values

### ISSUE THREAD

**Issue by kgriffs** [falconry/falcon#617]
When the client percent-encodes the commas for a value in the query string, the framework always treats that value as a blob, rather than a list of strings. This was originally done to support embedded documents, such as JSON, in the query string. What we would like is for this:

`things=id1%2Cid2%2Cid3`

...to be interpreted as a list by req.get_param_as_list, i.e.:

`['id1', 'id2', 'id3']

Whereas something like this:

`thing=%7B%22name%22%3A%20%22value%22%2C%20%22another%22%3A%20%22field%22%2C%20%22some%22%3A%20%22thing%22%7D`

...would be interpreted by req.get_param as:

```json
{"name": "value", "another": "field", "some": "thing"}
```

**Comment by ueg1990** [comment]
I was reading the source code about the get_param* methods....but i cannot seem to find where they are actually used? all i see is that they are used in the tests; that it. Am i missing something? Just trying to understand the internals of Falcon better

**Comment by yashmehrotra** [comment]
@kgriffs  Hi, I would like to work on this bug. I have understood the list part.

Also, I had a suggestion about the json part, how about adding a new function like `get_param_as_dict` or `get_param_as_json`.
Another way of doing it would be, using
```python
try:
    param = json.loads(body)
except ValueError:
    param = body
```

Which way do you suggest would be better ?

**Comment by ueg1990** [comment]
hey @yashmehrotra, can you share how and where the get_param* methods are used? i only see them being used in the tests. If you can help me figure out where in the source I can see it being used, i would be able to understand the source code better? :)

**Comment by yashmehrotra** [comment]
@ueg1990 There you go... https://github.com/falconry/falcon/blob/master/falcon/request.py#L713

**Comment by ueg1990** [comment]
@yashmehrotra i am aware of where these methods are defined. but where are they used/called?

**Comment by yashmehrotra** [comment]
They are called by the user, the one writing the web app in falcon

**Comment by gnomeby** [comment]
Yes in falcon 0.2 comma and escaped comma (%2C) worked in the same way. In 0.3 it is not true anymore. But it is not bad because:
When you are going to use comma as the part of value, you escape it. When you are going to use comma as the separator in value, you not escape it. In most cases you don't need to escape commas at all.

**Comment by kgriffs** [comment]
@yashmehrotra I think it would make sense to create a `get_param_as_dict()`. I broke this out into a separate issue (#789). Since a lot of users seem to be including JSON in their query strings, I think this is a relatively high priority.

The other work item was to improve detection of comma-separated lists. IMO this is lower priority. I broke it out into #790 and put it into the backlog.

BTW, I'll be at PyCon and will be sprinting on Falcon in case anyone would like to meet up. Thanks!
